### DISCUSSION TITLE
How to randomize subset of trials without any trial in the subset repeating?

### DISCUSSION THREAD

**Discussion by laurenbigelow** [jspsych/jsPsych#733] [Q&A] [answered]
Hi!

So I’m trying to randomize a subset of trials in my experiment (trials [2,4,7,10] in the code below), however I would like it to be that none of the trials repeat after they are run (which is what is happening now). So essentially what I want to happen in that the 3rd trial in the experiment could be either [2,4,7, or 10], but that once one of them is chosen, it won’t appear again in the 4th, 7th, or 10th trial. Does anyone know how I can do this?

Thanks!

Lauren

Here is the code for reference:

  sample: {
      type: 'alternate-groups',
      groups: [[0],[1],[2,4,7,10],[3],[2,4,7,10],[5],[6],[2,4,7,10],[8],[9],[2,4,7,10],[11]],
      randomize_group_order: false
    }

**Comment by vijaymarupudi** [comment]
You'll need some javascript.

Is this what you're looking for?

```javascript
var samples = [2, 4, 7, 10]

function popOneOff() {
  var rand_index = Math.round(Math.random() * (samples.length - 1))
  var item = samples[rand_index];
  samples.splice(rand_index, 1);
  return item;
}

var x = {
  type: "alternate-groups",
  groups: [
    [0],
    [1],
    [popOneOff()],
    [3],
    [popOneOff()],
    [5],
    [6],
    [popOneOff()],
    [8],
    [9],
    [popOneOff()],
    [11]
  ],
  randomize_group_order: false
};
```

**Comment by laurenbigelow** [comment]
Thanks for your help! I'm actually not familiar with javascript, so I'm not really sure how to put this into my code. (I removed my original code I had and put the code you suggested in but it didn't seem to change anything). If it helps, these are the trails that I'm using in the experiment. The numbers I listed for groups correspond to these trials, the first one being 0.

var test_procedure = {
    timeline: [likert_trial,  multi_select, text_page],
    timeline_variables: [
    { audio: 'sound/condition1/01.wav', trans: '<i><b>Allie: </b>When I was younger we had five kids in the </i>'+
       '<br><i>house which would be my two- three older brothers and then my sister and I.</i>', stim: '1'},
      { audio: 'sound/condition1/02.wav', trans: '<i><b>Courtney: </b>Ah, that’s kind of unusual for me. Usually if I’m </i>'+
      '<br><i>gonna drink a litre of anything it would be juice.</i>', stim: '2'},
      { audio: 'sound/condition1/03.wav', trans: '<i><b>Reggie: </b>Like we had to host games, like all the groomsmen </i>'+
      '<br><i>there.</i>' , stim: '3'},
      { audio: 'sound/condition1/04.wav', trans: '<i><b>James: </b>Basically, that’s the only move I made during my childhood </i>'+
      '<br><i>was just from one house to the house next-door on Alexander.</i>', stim: '4'},
      { audio: 'sound/condition1/05.wav', trans: '<i><b>Taylor: </b>Yeah and it was a- like a school welding competition. </i>'+
      '<br><i>We got smoked there, but. We had to build this big- ah for a pot, like a flower-pot. Just the cage.</i>', stim: '5'},
      { audio: 'sound/condition1/06.wav', trans: '<i><b>Ryan: </b> I think it was grade two actually. They had this Christmas advent calendar. </i>'+
      '<br><i>It’s one of those calendars with the chocolate things for each day.</i>', stim: '6'},
      { audio: 'sound/condition1/07.wav', trans: '<i><b>Mallory: </b>And I did like- I was involved in the sports, </i>'+
      '<br><i>I was involved in the bands, like, concert band, and.<i>', stim: '7'},
      { audio: 'sound/condition1/08.wav', trans: '<i><b>Justin: </b>The shop class is full of college students. Or </i>'+
      '<br><i>university students and, ah, he’s sitting- he was white as a ghost.</i>'+
      '<br><i>He’s sitting in the chair at the bottom.</i>', stim: '8'},
      { audio: 'sound/condition1/09.wav', trans: '<i><b>Katie: </b>Stacey from- one of the managers. He’s like my </i>'+
      '<br><i>favourite manager. Ever since that, anyway.</i>', stim: '9'},
      { audio: 'sound/condition1/10.wav', trans: '<i><b>Mark: </b> And I have, ah, one brother who’s a- about a year </i>'+
      '<br><i>younger than me. Ah, a sister who’s about three years younger than me.</i>', stim: '10'},
      { audio: 'sound/condition1/11.wav', trans: '<i><b>Brett: </b> So, at the starting the night when we were still at </i>'+
      '<br><i>the hotel, cause our hotel- the concert was at our hotel.</i>', stim: '11'},
      { audio: 'sound/condition1/12.wav', trans: '<i><b>Kyle: </b> My brother’s kinda just like laying about the house </i>'+
      '<br><i>right now, waiting for college, I guess. He really doesn’t really </i>'+
      '<br><i>seem to do much when he’s not at school.</i>', stim: '12'}
    ],
};
timeline.push(test_procedure);

**Comment by vijaymarupudi** [comment, marked as answer]
Are you looking for something like this?

```javascript
var test_procedure = {
  timeline: [likert_trial, multi_select, text_page],
  sample: {
    type: "custom",
    fn: function(samples) {
      function popOneOff() {
        var rand_index = Math.round(Math.random() * (samples.length - 1));
        var item = samples[rand_index];
        samples.splice(rand_index, 1);
        return item;
      }

      return [
        0,
        1,
        popOneOff(),
        3,
        popOneOff(),
        5,
        6,
        popOneOff(),
        8,
        9,
        popOneOff(),
        11
      ];
    }
  },
  timeline_variables: [
    {
      audio: "sound/condition1/01.wav",
      trans:
        "Allie: When I was younger we had five kids in the " +
        " house which would be my two- three older brothers and then my sister and I.",
      stim: "1"
    },
    {
      audio: "sound/condition1/02.wav",
      trans:
        "Courtney: Ah, that’s kind of unusual for me. Usually if I’m " +
        " gonna drink a litre of anything it would be juice.",
      stim: "2"
    },
    {
      audio: "sound/condition1/03.wav",
      trans:
        "Reggie: Like we had to host games, like all the groomsmen " +
        " there.",
      stim: "3"
    },
    {
      audio: "sound/condition1/04.wav",
      trans:
        "James: Basically, that’s the only move I made during my childhood " +
        " was just from one house to the house next-door on Alexander.",
      stim: "4"
    },
    {
      audio: "sound/condition1/05.wav",
      trans:
        "Taylor: Yeah and it was a- like a school welding competition. " +
        " We got smoked there, but. We had to build this big- ah for a pot, like a flower-pot. Just the cage.",
      stim: "5"
    },
    {
      audio: "sound/condition1/06.wav",
      trans:
        "Ryan: I think it was grade two actually. They had this Christmas advent calendar. " +
        " It’s one of those calendars with the chocolate things for each day.",
      stim: "6"
    },
    {
      audio: "sound/condition1/07.wav",
      trans:
        "Mallory: And I did like- I was involved in the sports, " +
        " I was involved in the bands, like, concert band, and.",
      stim: "7"
    },
    {
      audio: "sound/condition1/08.wav",
      trans:
        "Justin: The shop class is full of college students. Or " +
        " university students and, ah, he’s sitting- he was white as a ghost." +
        " He’s sitting in the chair at the bottom.",
      stim: "8"
    },
    {
      audio: "sound/condition1/09.wav",
      trans:
        "Katie: Stacey from- one of the managers. He’s like my " +
        " favourite manager. Ever since that, anyway.",
      stim: "9"
    },
    {
      audio: "sound/condition1/10.wav",
      trans:
        "Mark: And I have, ah, one brother who’s a- about a year " +
        " younger than me. Ah, a sister who’s about three years younger than me.",
      stim: "10"
    },
    {
      audio: "sound/condition1/11.wav",
      trans:
        "Brett: So, at the starting the night when we were still at " +
        " the hotel, cause our hotel- the concert was at our hotel.",
      stim: "11"
    },
    {
      audio: "sound/condition1/12.wav",
      trans:
        "Kyle: My brother’s kinda just like laying about the house " +
        " right now, waiting for college, I guess. He really doesn’t really " +
        " seem to do much when he’s not at school.",
      stim: "12"
    }
  ]
};
timeline.push(test_procedure);

```

**Reply by jodeleeuw** [reply]
I think this is close, but `samples` will be an array of all the trial indexes in the timeline variable list.

I think an alternative to the custom sampling function you describe is:

```js
fn: function() {
	var trials_to_randomize = [2,4,7,10];
	var random_order = jsPsych.randomization.shuffle(trials_to_randomize);

      	return [
        	0,
        	1,
		random_order[0],   
        	3,
        	random_order[1], 
        	5,
        	6,
        	random_order[2], 
        	8,
        	9,
        	random_order[3], 
        	11
      	];
}
```

**Reply by jodeleeuw** [reply]
For completeness, here's the entire code chunk

```js
var test_procedure = {
  timeline: [likert_trial, multi_select, text_page],
  sample: {
    type: "custom",
    fn: function() {
	var trials_to_randomize = [2,4,7,10];
	var random_order = jsPsych.randomization.shuffle(trials_to_randomize);

      	return [
        	0,
        	1,
		random_order[0],   
        	3,
        	random_order[1], 
        	5,
        	6,
        	random_order[2], 
        	8,
        	9,
        	random_order[3], 
        	11
      	];
    }
  },
  timeline_variables: [
    {
      audio: "sound/condition1/01.wav",
      trans:
        "Allie: When I was younger we had five kids in the " +
        " house which would be my two- three older brothers and then my sister and I.",
      stim: "1"
    },
    {
      audio: "sound/condition1/02.wav",
      trans:
        "Courtney: Ah, that’s kind of unusual for me. Usually if I’m " +
        " gonna drink a litre of anything it would be juice.",
      stim: "2"
    },
    {
      audio: "sound/condition1/03.wav",
      trans:
        "Reggie: Like we had to host games, like all the groomsmen " +
        " there.",
      stim: "3"
    },
    {
      audio: "sound/condition1/04.wav",
      trans:
        "James: Basically, that’s the only move I made during my childhood " +
        " was just from one house to the house next-door on Alexander.",
      stim: "4"
    },
    {
      audio: "sound/condition1/05.wav",
      trans:
        "Taylor: Yeah and it was a- like a school welding competition. " +
        " We got smoked there, but. We had to build this big- ah for a pot, like a flower-pot. Just the cage.",
      stim: "5"
    },
    {
      audio: "sound/condition1/06.wav",
      trans:
        "Ryan: I think it was grade two actually. They had this Christmas advent calendar. " +
        " It’s one of those calendars with the chocolate things for each day.",
      stim: "6"
    },
    {
      audio: "sound/condition1/07.wav",
      trans:
        "Mallory: And I did like- I was involved in the sports, " +
        " I was involved in the bands, like, concert band, and.",
      stim: "7"
    },
    {
      audio: "sound/condition1/08.wav",
      trans:
        "Justin: The shop class is full of college students. Or " +
        " university students and, ah, he’s sitting- he was white as a ghost." +
        " He’s sitting in the chair at the bottom.",
      stim: "8"
    },
    {
      audio: "sound/condition1/09.wav",
      trans:
        "Katie: Stacey from- one of the managers. He’s like my " +
        " favourite manager. Ever since that, anyway.",
      stim: "9"
    },
    {
      audio: "sound/condition1/10.wav",
      trans:
        "Mark: And I have, ah, one brother who’s a- about a year " +
        " younger than me. Ah, a sister who’s about three years younger than me.",
      stim: "10"
    },
    {
      audio: "sound/condition1/11.wav",
      trans:
        "Brett: So, at the starting the night when we were still at " +
        " the hotel, cause our hotel- the concert was at our hotel.",
      stim: "11"
    },
    {
      audio: "sound/condition1/12.wav",
      trans:
        "Kyle: My brother’s kinda just like laying about the house " +
        " right now, waiting for college, I guess. He really doesn’t really " +
        " seem to do much when he’s not at school.",
      stim: "12"
    }
  ]
};
timeline.push(test_procedure);
```

**Reply by vijaymarupudi** [reply]
Oh you're right, made a mistake when translating to the sample function! Use @jodeleeuw's answer, @laurenbigelow.

**Reply by laurenbigelow** [reply]
Yes, that's just what I needed! Thanks both of you for the help!
